tests/integration/test_bundle.py

```diff
@@ -5,21 +5,97 @@
     async with ops_test.fast_forward():
         await ops_test.model.deploy("./releases/latest/mysql-k8s-bundle.yaml", trust=True)
 
-        await ops_test.model.wait_for_idle(
-            apps=[MYSQL_APP, TLS_APP], status="active", timeout=5 * 60
-        )
+        await ops_test.model.wait_for_idle(apps=[MYSQL_APP], status="active", timeout=5 * 60)
 
         await ops_test.model.wait_for_idle(apps=[ROUTER_APP], status="waiting", timeout=5 * 60)
+
+
+async def test_mysql_primary_switchover(ops_test: OpsTest):
+    """Ensures writes continue after the primary is killed."""
+    # Deploy application & relate to router
+    application_charm = await ops_test.build_charm("./tests/integration/application_charm/")
+    await ops_test.model.deploy(application_charm, application_name=APPLICATION_APP, num_units=1)
+    await ops_test.model.relate(f"{APPLICATION_APP}:database", f"{ROUTER_APP}:database")
+    await ops_test.model.wait_for_idle(
+        apps=[MYSQL_APP, ROUTER_APP, APPLICATION_APP],
+        status="active",
+        raise_on_blocked=True,
+        timeout=15 * 60,
+    )
+    # Start writes
+    await ensure_all_units_continuous_writes_incrementing(ops_test)
+    # Kill primary mysql
+    mysql_units = ops_test.model.applications[MYSQL_APP].units
+    primary = await get_primary_unit(ops_test, mysql_units[0], MYSQL_APP)
+    # ensure all units in the cluster are online
+    assert await ensure_n_online_mysql_members(
+        ops_test, 3
+    ), "The deployed mysql application is not fully online"
+
+    mysql_pid = await get_process_pid(
+        ops_test, primary.name, MYSQL_CONTAINER_NAME, MYSQLD_PROCESS_NAME
+    )
+
+    await send_signal_to_pod_container_process(
+        ops_test,
+        primary.name,
+        MYSQL_CONTAINER_NAME,
+        MYSQLD_PROCESS_NAME,
+        "SIGKILL",
+    )
+
+    # Wait for the SIGKILL above to take effect before continuing with test checks
+    time.sleep(10)
+
+    assert await ensure_n_online_mysql_members(
+        ops_test, 3
+    ), "The mysql application is not fully online after sending SIGKILL to primary"
+
+    # ensure that the mysqld process got restarted and has a new process id
+    new_mysql_pid = await get_process_pid(
+        ops_test, primary.name, MYSQL_CONTAINER_NAME, MYSQLD_PROCESS_NAME
+    )
+    assert (
+        mysql_pid != new_mysql_pid
+    ), "The mysql process id is the same after sending it a SIGKILL"
+
+    new_primary = await get_primary_unit(ops_test, mysql_units[0], MYSQL_APP)
+    assert (
+        primary.name != new_primary.name
+    ), "The mysql primary has not been reelected after sending a SIGKILL"
+
+    # Ensure writes continue
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(
+            apps=[MYSQL_APP, ROUTER_APP, APPLICATION_APP],
+            status="active",
+            raise_on_blocked=True,
+            timeout=15 * 60,
+        )
+        await ensure_all_units_continuous_writes_incrementing(ops_test)
```

Comment:
i think that we also need to query the max value in the database before sending a SIGKILL and to then ensure that the value has increased right after the re-election (we need to see that there is no interruption in service). a more robust way to test this would be to "freeze" continuous writes if there is even a single query that is interrupted when we send a SIGKILL to the primary process

Reply:
yes, i believe this approach works. there are more intricate ways of ensuring that mysql-router does not interrupt service when a primary dies, but this is good (we can further refine the continuous writes application to detect service interruption and do something to flag it in a future PR)